dotnet/src/InternalUtilities/src/Text/StreamJsonParser.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.Json.Nodes;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.SemanticKernel.Text;
+
+#pragma warning disable CA1812 // Internal class that is apparently never instantiated
+#pragma warning disable CA1846 // Prefer 'AsSpan' over 'Substring' when span-based overloads are available
+
+/// <summary>
+/// Internal class for parsing a stream of text which contains a series of discrete JSON strings into en enumerable containing each separate JSON string.
+/// </summary>
+/// <remarks>
+/// This class is thread-safe.
+/// </remarks>
+internal sealed class StreamJsonParser
+{
+    /// <summary>
+    /// Parses a Stream containing JSON data and yields the individual JSON objects.
+    /// </summary>
+    /// <param name="stream">The Stream containing the JSON data.</param>
+    /// <param name="validateJson">Set to true to enable checking json chunks are well-formed. Default is false.</param>
```

Comment:
What type of inputs does this method accept when `validateJson` is set to false? As far as I can tell it is implementing its own JSON parser, so clearly some structure is necessary for this to identify the end of the next element.

Reply:
Put differently, would exposing some lax parsing mode in `Utf8JsonReader` remove the need for you to implement a bespoke parser?

Reply:
StreamJsonparser only counts { and } and returns string when counter will be 0. validateJson = true additionally invoke JsonNode.Parse after completing json string.

Reply:
I tried use Utf8JsonReader in early version of parser but often getting parser exceptions.

Reply:
> I tried use Utf8JsonReader in early version of parser but often getting parser exceptions.

Can you share examples of JSON data that is causing `Utf8JsonReader` to fail?

Reply:
> StreamJsonparser only counts { and } and returns string when counter will be 0. validateJson = true additionally invoke JsonNode.Parse after completing json string.

That might be a quick way to split documents, but given that you expect the input data to be invalid JSON, one misplaced curly brace could result in erroneous groupings spilling over to the entire data stream. cc @grabyourpitchforks

Reply:
@eiriktsarpalis I don't remember exactly, I can't find first version of parser in my git history. But probably exception from Utf8JsonReader occurred if current slice contained small amount of data like `,{ "d` then was throwing exception not recognized TokenType, but I am not sure now how it was...

Not exactly invalid input, this Parser is designed to ignore SSE tags and return only json data. And for example if input will contain `data: } { "person": { "name": "example" } }` exception will be thrown

Reply:
> But probably exception from Utf8JsonReader occurred if current slice contained small amount of data like ,{ "d then was throwing exception not recognized TokenType, but I am not sure now how it was...

That particular case suggests that `Utf8JsonReader` wasn't configured to support partial data (this can be done via the [`isFinalBlock` constructor parameter](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonreader.-ctor?view=net-8.0#system-text-json-utf8jsonreader-ctor(system-buffers-readonlysequence((system-byte))-system-boolean-system-text-json-jsonreaderstate))). The second example is blatantly invalid JSON and `Utf8JsonReader` won't accept it.

Reply:
I had isFinalBlock set to false, for sure, but I will try to reproduce problematic code. I'll get back to you later.

Reply:
@eiriktsarpalis @stephentoub okay I can't reproduce same code... So it's probably that was any my fault exception was throwing when parsing but JsonDocument parser still doesn't solve from with SSE input where data contains other data than json. That's why my parser is more complicated, implementation from docs also is not so simple https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-utf8jsonreader#read-from-a-stream-using-utf8jsonreader

Reply:
What's the situation where non-json support is needed? And isn't this still based on looking for bonds beginning/ending with braces?